src/Confluent.Kafka.Avro/SpecificDeserializerImpl.cs

```diff
@@ -0,0 +1,139 @@
+            {
+                ReaderSchema = Avro.Schema.Parse("string");
+            }
+            else if (typeof(T).Equals(typeof(float)))
+            {
+                ReaderSchema = Avro.Schema.Parse("float");
+            }
+            else if (typeof(T).Equals(typeof(long)))
+            {
+                ReaderSchema = Avro.Schema.Parse("long");
+            }
+            else if (typeof(T).Equals(typeof(byte[])))
+            {
+                ReaderSchema = Avro.Schema.Parse("bytes");
+            }
+            else
+            {
+                throw new ArgumentException(
+                    $"{nameof(AvroDeserializer<T>)} " +
+                    "only accepts type parameters of int, bool, double, string, float, " +
+                    "long, byte[], instances of ISpecificRecord and subclasses of SpecificFixed."
+                );
+            }
+        }
+
+        public T Deserialize(string topic, byte[] array)
+        {
+            // Note: topic is not necessary for deserialization (or knowing if it's a key 
+            // or value) only the schema id is needed.
+
+            using (var stream = new MemoryStream(array))
+            using (var reader = new BinaryReader(stream))
+            {
+                var magicByte = reader.ReadByte();
+                if (magicByte != Constants.MagicByte)
+                {
+                    // may change in the future.
+                    throw new InvalidDataException($"magic byte should be 0, not {magicByte}");
+                }
+                var writerIdBigEndian = reader.ReadInt32();
+                var writerId = IPAddress.NetworkToHostOrder(writerIdBigEndian);
+
+                DatumReader<T> datumReader = null;
+
+                lock (datumReaderLock)
+                {
+                    datumReaderBySchemaId.TryGetValue(writerId, out datumReader);
+                    if (datumReader == null)
+                    {
+                        if (datumReaderBySchemaId.Count > schemaRegistryClient.MaxCachedSchemas)
+                        {
+                            datumReaderBySchemaId.Clear();
+                        }
+
+                        // it's a good idea to retain the lock over this blocking call since there
+                        // may be concurrent deserialize calls for the same T, and in that case it's
+                        // best not to hit Schema Registry more than once for the same information.
+                        var writerSchemaJson = schemaRegistryClient.GetSchemaAsync(writerId).Result;
+                        var writerSchema = Avro.Schema.Parse(writerSchemaJson);
+
+                        datumReader = new SpecificReader<T>(writerSchema, ReaderSchema);
+                        datumReaderBySchemaId[writerId] = datumReader;
+                    }
+                }
+
+                return datumReader.Read(default(T), new BinaryDecoder(stream));
+            }
+        }
```

Comment:
this code looks almost identical to GenericDeserializerImpl, maybe refactor into one common base class?

Reply:
you're right, though the code is relatively short, I think annoying to refactor (since the difference is in typing), and introducing a more complex structure (especially one that includes a base class) comes with the downside in making the code more difficult to follow (refer to the java implementation!). Also "rule of three" https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

Reply:
(so leaving as is)